qbraid_qir/qasm3/visitor.py

```diff
@@ -133,14 +139,26 @@ def entry_point(self) -> str:
     def finalize(self) -> None:
         self._builder.ret(None)
 
-    def _set_scope(self, scope: Scope) -> None:
-        self._scope = scope
+    def _push_scope(self, scope: dict) -> None:
+        self._scope.append(scope)
+
+    def _pop_scope(self) -> None:
+        assert len(self._scope) >= 1
+        self._scope.pop()
+
+    def _get_scope(self) -> dict:
+        assert len(self._scope) >= 1
+        return self._scope[-1]
+
+    def _check_in_scope(self, var_name: str) -> bool:
+        curr_scope = self._get_scope()
+        return var_name in curr_scope
 
-    def _in_gate(self) -> bool:
-        return self._scope == Scope.GATE
+    def _update_scope(self, variable: Variable) -> None:
+        self._scope[-1][variable.name] = variable
 
     def _in_global_scope(self) -> bool:
-        return self._scope == Scope.GLOBAL and self._context == Context.GLOBAL
+        return len(self._scope) == 1 and self._context == Context.GLOBAL
 
     def _in_function(self) -> bool:
         return self._scope == Scope.FUNCTION
```

Comment:
Yes, I thought of different ways to incorporate scope into the code. After referring to the openqasm source, it made sense to define where we are using the scope size - 
1. If `len(scope) == 1`, it is global
2. If `len(scope) > 1`, it is some function 

I don't know how the `_in_function` method will be incorporated, but I felt this would probably be needed when adding support for the subroutines.